mkchkimg: use higher version code

jclehner opened the pull request:
This patch changes the version code of the image header
from `1.1.99_0.0.0.0` to `99.99.99_99.99.99.99`. This
is neccessary on some devices where the stock firmware
checks the version field, possibly preventing third-party
firmware from being installed.

Signed-off-by: Joseph C. Lehner <joseph.c.lehner@gmail.com>

ynezz commented:
@jclehner Why do you need this change? Your commit message should probably include this detail.

jclehner commented:
Updated the commit message. I don't actually own a Netgear R7000, but the image building code for that device should probably make use of this change. Alternatively, one might consider changing the default image version used by `mkchkimg` to something higher than `1.1.99.0_0.0.0`.

ynezz commented:
@f00b4r0 Hi, I'm wondering if you could by any chance kindly review this and ideally test it as well. Thanks!

f00b4r0 reviewed (commented):
Here's a quick review based on code-read only.

I personally find the changes a bit invasive (especially the - IMO unnecessary - renaming of the `reserved` field), and I believe the commit message will require a bit more explanation as to why this is necessary and what are the implications (i.e. be a bit more specific as to the reverse-engineered behavior of the bootloader).

In particular, I wonder if the same version number can be be reflashed on the concerned hardware (I assume it can). In that case it might be just simpler to arbitrarily set the fields to e.g. 99.99.99 or even 255.255.255 if we don't require this field to bear a specific meaning (I believe this was the original intent of the code).

This way we can do away with the extra logic and avoid having to play cat and mouse with OEM numbering.

HTH

**Edit**: it seems #1792 does exactly what I suggest above... Although I suppose one downside of this approach is that it probably makes it impossible to revert to stock firmware with the new bootloader behavior. If this PR goes through, I suspect it's going to need some documenting effort and put some red flags where appropriate. HTH.

jclehner committed:
mkchkimg: use higher version code

This patch changes the version code of the image header
from `1.1.99_0.0.0.0` to `99.99.99_99.99.99.99`. This
is neccessary on some devices where the stock firmware
checks the version field, possibly preventing third-party
firmware from being installed.

Signed-off-by: Joseph C. Lehner <joseph.c.lehner@gmail.com>

jclehner commented:
I've updated the patch and removed the version argument alltogether. It now uses version `99.99.99_99.99.99.99`, which should prevent any collision with stock images.

ynezz commented:
Funny, I've just found out about https://github.com/openwrt/openwrt/pull/1792

nharward commented:
Verifying this patch works for an R8000 purchased on 2020-04-16 against the 1.0.4.46 stock firmware - nothing else I tried or could find worked. After running the steps below I was able to use `github.com:jclehner/nmrpflash` (master, commit e95526d06cd4d8869befb79a26ade8bbc2262265) to flash back to the 1.0.4.46 stock firmware, then repeat the process a second time successfully.

Steps, running (a docker container of) `Debian 10 (buster)`:
1. Install all dependencies listed at https://openwrt.org/docs/guide-developer/build-system/install-buildsystem, including Subversion and Mercurial mentioned at the top
1. Clone source, check out the v19.07.02 tag
1. Apply patch from this PR
1. `make prereq` # customized in menuconfig to use `wpad-mesh-openssl` and remove LuCI 
1. `make` # produces `build_dir/target-arm_cortex-a9_musl_eabi/linux-bcm53xx/tmp/openwrt-bcm53xx-netgear-r8000-squashfs.chk`
1. Upload artifact via stock firmware's web interface
1. Router rebooted normally as part of the update; for good measure then powered the router off for 10 seconds then back on again

NicolaF commented:
Hi,

I just ran into the same issue with a brand new R8000. I downgraded the stock firmware as much as I could, but to no avail. I could not downgrade below 1.0.4.12_10.1.46, and it didn't help.

Applying the proposed patch did indeed trick the stock firmware.

f00b4r0 commented:
@NicolaF @nharward Can any of you confirm that it is possible to restore stock firmware after flashing OpenWRT with this patch? If so, @ynezz I think the current proposal can be merged as is.

ynezz commented:
@f00b4r0 Ok, thanks, can I add your `Reviewed-by` or such?

f00b4r0 commented:
@ynezz Sure, if you want :)

nharward commented:
> @NicolaF @nharward Can any of you confirm that it is possible to restore stock firmware after flashing OpenWRT with this patch? If so, @ynezz I think the current proposal can be merged as is.

Yes I can confirm flashing back to stock using the excellent `github.com:jclehner/nmrpflash` utility. I would up discovering the R8000 doesn't have the hardware to support my 80211s use case (d'oh!) and had to flash it back to stock to give to someone else - works fine for them.

ynezz commented:
@nharward thank you for checking

patricius972 commented:
I am running into the same problem with a R7000, the chk file is rejected at tftp. Any reason this patch is not merged yet? I am building an test image to try it.

ynezz commented:
@patricius972 just pushed in 0a73c61cb933daad40309c22d8eea6034454dcbd, sorry for the delay :)

patricius972 commented:
Thanks, my test image worked. that would avoid going the serial port for some other.
